### PR TITLE
pull request to add first spotify commands

### PR DESCRIPTION
**Pull request by kat-kel**
This is a first draft at creating a new command for minet, `spotify`, which queries Spotify's Web API.

It requires a new `"spotify"` section in the rcfile with what Spotify calls a "client_id" and "client_secret". [They're generated when you create an app for the API on a Spotify account](https://developer.spotify.com/documentation/web-api/tutorials/getting-started), which I did for testing. I can give you my testing tokens if you want. They can also be passed individually to the command as options.
```json
{"spotify": 
    {
        "client_id": "AAAA",
        "client_secret": "BBBB"
    }
}
```

The first commands I've proposed to add are [general searches](https://developer.spotify.com/documentation/web-api/reference/search) on the API of (1) podcast shows and (2) podcast episodes, based on a keyword match in the show or episode's name or description. In addition to a keyword (query), the search can take on a parameter that filters the country (`--market`) in which the podcast was distributed.

For example,
```shell
>>> minet spotify episodes "reforme retraites" --market FR --rcfile config.yml -o reforme_episodes.csv
```

One thing I don't like yet is that if you give an invalid country code to the `--market` option (i.e. `--market FRA`), the `request` function used in `minet/spotify/api_client.py` throws a `KeyError` and it's not descriptive of the problem.

### CODE REVIEW

**Comment by kat-kel** [comment]
One particularity about Spotify's API is that an application does not have a permanent access token. Every 30 minutes, you need to request a new one. My solution was to use minet's `request` with the "POST" method, see below, to get a new access token each time the API Wrapper is instantiated:
https://github.com/medialab/minet/blob/44a9aef972a68c72d9b5c3805b121fffee1f7e0f/minet/spotify/api_client.py#L14-L34

File: minet/spotify/api_client.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        self.retryer = create_request_retryer()
+        self.cli_search_kwargs = kwargs
+
+    @retrying_method()
+    def request_json(self, url):
+        response = request(
+            url=url, headers={"Authorization": f"Bearer {self.access_token}"}
+        )
+        if response.status == 429:
+            time.sleep(30)
```
Comment (Yomguithereal) [inline minet/spotify/api_client.py:63]: We should log this probably at one point. But let's discuss this later.

File: minet/spotify/api_client.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        url = forge_search_url(search_kwargs)
+        return self.generator(url, SpotifyShow.from_payload)
+
+    def episodes(self, query):
+        search_kwargs = {"q": query, "type": "episode"}
+        search_kwargs.update(self.cli_search_kwargs)
+        url = forge_search_url(search_kwargs)
+        return self.generator(url, SpotifyShowEpisode.from_payload)
+
+    def generator(self, url, formatter):
```
Comment (Yomguithereal) [inline minet/spotify/api_client.py:84]: This should probably named `paginate` or something like this?

File: minet/spotify/api_client.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        go_to_next_page = True
+        while go_to_next_page:
```
Comment (Yomguithereal) [inline minet/spotify/api_client.py:86]: `while True` with a `break` somewhere feels clearer to me.

File: minet/spotify/api_client.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            result = self.request_json(url)
+            # If the result doesn't declare a type
+            if result.get("items"):
+                data = result
+            # Or if the result declares the media type
+            elif len(list(result.keys())) == 1:
+                type = list(result.keys())[0]
+                data = result[type]
+            else:
+                raise KeyError
```
Comment (Yomguithereal) [inline minet/spotify/api_client.py:96]: This should be a custom error somehow.

File: minet/spotify/api_client.py (hunk 2 above)
Comment (Yomguithereal) [reply minet/spotify/api_client.py:84]: Note that I would probably make this handle only the actual API pagination and yield the payload but maybe not extract the items if this is different for each kind of call.

**Review by Yomguithereal — CHANGES_REQUESTED** [review]

**Pull request closed without merging**